diff --git a/aimet_zoo_torch/ffnet/model/ffnet_S_mobile.py b/aimet_zoo_torch/ffnet/model/ffnet_S_mobile.py
index 8935a5d..ef05033 100644
--- a/aimet_zoo_torch/ffnet/model/ffnet_S_mobile.py
+++ b/aimet_zoo_torch/ffnet/model/ffnet_S_mobile.py
@@ -88,10 +88,7 @@ def segmentation_ffnet40S_dBBB_mobile():
         model_name="ffnnet40S_dBBB_mobile",
         backbone=resnet.Resnet40S_D,
         pre_downsampling=False,
-        pretrained_weights_path=os.path.join(
-            model_weights_base_path,
-            "ffnet40S/ffnet40S_dBBB_cityscapes_state_dict_quarts.pth",
-        ),
+        pretrained_weights_path="ffnet40S_dBBB_cityscapes_state_dict_quarts.pth",
         strict_loading=True,
     )
 
