diff --git a/yolov6/utils/general.py b/yolov6/utils/general.py
index cb4418c..5300651 100644
--- a/yolov6/utils/general.py
+++ b/yolov6/utils/general.py
@@ -5,7 +5,7 @@ import glob
 import math
 import torch
 import requests
-import pkg_resources as pkg
+from packaging import version
 from pathlib import Path
 from yolov6.utils.events import LOGGER
 
@@ -119,7 +119,7 @@ def check_img_size(imgsz, s=32, floor=0):
 
 def check_version(current='0.0.0', minimum='0.0.0', name='version ', pinned=False, hard=False, verbose=False):
     # Check whether the package's version is match the required version.
-    current, minimum = (pkg.parse_version(x) for x in (current, minimum))
+    current, minimum = (version.parse(x) for x in (current, minimum))
     result = (current == minimum) if pinned else (current >= minimum)  # bool
     if hard:
         info = f'⚠️ {name}{minimum} is required by YOLOv6, but {name}{current} is currently installed'
